design/node-agent-concurrency.md

```diff
@@ -0,0 +1,78 @@
+- Create a mechanism for users to specify the concurrent number of VGDP per node
+
+## Non-Goals
+- VGDP instances from different nodes always run in concurrent since in most common cases the resources are isolated. For special cases that some resources are shared across nodes, there is no support at present
+- In practice, restores run in prioritized scenarios, e.g., disarster recovery. However, the current design doesn't consider this difference, a VGDP instance for a restore is blocked if it reaches to the limit of the concurrency, even though the ones block it are for backups. If users do meet some problems here, they should consider to stop the backups first
+- Sometimes, users wants to block VGDP instances from running in a specific node, this is not considered in the current design. To archive this, more modules need to be considered, just setting VGDP instances' concurrent number to 0 won't bring the ideal reasult
+
+## Solution
+
+### Global concurrent number
+We add a node-agent server parameter ```data-path-concurrent-num``` to accept a concurrent number that will be applied to all nodes if the per-node number is not specified.  
+The number starts from 1 which means there is no concurrency, only one instance of VGDP is allowed. There is no roof limit.    
+If the node-agent server parameter is not specified, a hard-coded default value will be used, the value is set to 1.   
+
+### Per-node concurrent number
+We allow users to specify different concurrent number per node, for example, users can allow 3 concurrent instances in Node-1, 2 instances in Node-2 and 1 instance in Node-3. This is for below considerations:
+- The resources may be different among nodes. Then users could specify smaller concurrent number for nodes with less resources while larger number for the ones with more resources
+- Help users to isolate critical environments. Users may run some critical workloads in some specified nodes, since VGDP instances may take large resource consumption, users may only want to run them in the other nodes or run less number of instances in the nodes with critical workloads
+
+The range of Per-node concurrent number is the same with Global concurrent number.  
+Per-node concurrent number is preferable to Global concurrent number, so it will overwrite the Global concurrent number for that node.  
+Per-node concurrent number is implemented in a separate configMap named ```node-agent-configs```. This configMap is not created by Velero, users should create it manually on demand. The configMap should be created before node-agent server starts, otherwise, node-agent server should be restarted in order to make the effect.  
+
+The ```node-agent-configs``` configMap may be used for other purpose of configuring node-agent in future, at present, there is only one kind of data ```data-path-concurrency```. Its data struct is like below: 
+
+```go
+type DataPathConcurrency struct {
```

Comment:
Was it considered to make this a CRD singleton cluster scoped CR?

Reply:
Not sure if I get your question, let me try to answer. By the design, this is singleton for a single Velero installation. If there are more than one installation in a cluster, I didn't consider it for the design, but technically, yes, all the Velero installations will follow the same cm.

Reply:
So either, I think this should be a cluster-scoped resource as a CRD (not a config map). I think it will be really confusing otherwise.

Reply:
As far as my understanding, this per-node number is a static configuration, so I don't understand the benefit to make it a CRD.
If you are talking about dynamically watching it, as [mentioned here](https://github.com/vmware-tanzu/velero/pull/6950#discussion_r1373986044), I don't think it is necessary to dynamically watch a configuration that is not to be frequently changed.

Reply:
It may not be frequently changed, BUT when you change it, what is the process now, and how does it affect running backups?

Reply:
@shawn-hurley any backup or restore in an InProgress state will be marked failed and a new one will need to be created, since we have not yet solved the problem of making InProgress  backups or restores resumable.

Reply:
First of all, it should not affect the running backup, because you cannot stop one VGDP instance if it is already started, e.g., previously, the backup gets 3 concurrency, now if we change it to 2, nothing we can do to the backup.

Secondly, the current logic after the number is changed is that users have to restart the node-agent server when there is no running backups, otherwise, the update never takes effect.
We are doing in this way is just to prevent users frequently change this number and we don't want the already submitted backups/restores get affected.

Reply:
But wait, we're talking about the node agent here. That won't fail backups or restores directly, but it *will* cause in-progress DataUploads or DataDownloads to be canceled. I'm assuming it will fail PodVolumeBackups/PodVolumeRestores as well.

Reply:
So both the global number change and per-node number change won't take effect unless node-agent server restarts and node-agent server is only recommended to restart when there are no running backups/restores

Reply:
I am worried about the brittleness of this with a backup and restore system.

Reply:
1. For the already running backups/restores, I do believe this is a reasonable behavior, we cannot peel a data path thread (which for example has backed up tones of data) whenever the number is changed by users.
2. For the backups/restores to be created/submitted, we can use the new number without restarting node-agent server, but I don't think it is something must have, especially in the first version. 

For 2, I am open for a vote of this. It is just efforts/complexity vs. user friendliness

Reply:
I don't think I am following 1 unless you are spelling out the agreement on letting running threads continue running and reducing the number over time.

I believe that if you implemented the system described in 1 and 2 it would be more resilient and have a better UX (especially when you say need to bump up/down these resources for a $REASON.

Reply:
> the agreement on letting running threads continue running and reducing the number over time

Yes, this is what I meant, the limit should only work on the new coming VGDP instance creations.